### ISSUE TITLE
Declaration order confuses CFE's notion of sealed when mixins involved

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/sdk#52048]
The following program is accepted by both the CFE and the analyzer:
```dart
main() {
  f(B());
}

f(A a) {
  switch (a) {
    case B():
      print("It's a B");
  }
}

sealed class A {}

class B extends A with M {
  B();
}

mixin M {}
```
However, if you swap the order of declaration of classes `A` and `B`, i.e.:
```dart
main() {
  f(B());
}

f(A a) {
  switch (a) {
    case B():
      print("It's a B");
  }
}

class B extends A with M {
  B();
}

sealed class A {}

mixin M {}
```
Then you get the following error from the CFE:
```
../../tmp/proj/test.dart:6:11: Error: The type 'A' is not exhaustively matched by the switch cases since it doesn't match '_B&A&M()'.
 - 'A' is from '../../tmp/proj/test.dart'.
Try adding a default case or cases that match '_B&A&M()'.
  switch (a) {
          ^
```
Reproduced as of 90d84c849472b375e0fbb7321f800208180cd14b.

If we can fix this before the release, I suspect it will be worth cherry-picking to stable.
CC @johnniwinther

**Comment by leafpetersen** [comment]
If this is just restricted to mixing in things onto a sealed class (why do we even have that lever?) I think I'm on the fence about cherry picking, it doesn't feel likely to be a widespread problem.  Definitely worth fixing though, and we can then consider.  Separately, the error message there is pretty terrible.

**Comment by stereotype441** [comment]
@leafpetersen 
> If this is just restricted to mixing in things onto a sealed class (why do we even have that lever?)...

I guess our intuitions disagree about how useful it is to do that.  That's fine--to each their own.

But for what it's worth, I didn't find this bug by deliberately poking at edge cases.  I found it because I was trying out patterns in a side project over the weekend, and it arose naturally from a genuine attempt to write a useful program 😃

(The details of what I was trying to do, if you're curious: I had a sealed class with several derived classes, and I had a mixin (in a different library) that I could use to add some extra debugging functionality to a class.  I wanted to add the extra debugging functionality to some of the derived classes but not all of them.  I was using the standard "sort declarations" feature in the IDE to keep my file organized, and so I happened to hit on the sort order that provoked the bug.)
